File: src/os-speedtest/src/usr/local/opnsense/scripts/speedtest/config_mirror.php
#!/usr/local/bin/php
<?php
/* Move one section between config.xml and one JSON object, understanding
   neither. This script exists because only PHP can read and write an OPNsense
   configuration the way OPNsense does; every question of what a field means,
   which value is sane and what to do about a missing one is answered in
   config_mirror.py, so that it can be answered under test. Adding knowledge
   here would put policy somewhere no test can reach it. */
require_once('script/load_phalcon.php');

use OPNsense\Core\Config;
use OPNsense\Speedtest\Backup;

/* The same address as <mount> in models/OPNsense/Speedtest/Backup.xml. The
   suite compares the two, so moving the model cannot silently orphan the
   section a restored configuration still carries. */
function backup_lock_current(Config $config): void
{
    $config->lock();
    /* Core's stream reader can downgrade the lock during reload. Reacquire
       exclusive mode and refresh the mutable document without another reader. */
    $config->lock(false);
    $settings = new \OPNsense\Core\AppConfig();
    $fresh = simplexml_load_string(
        file_get_contents($settings->application->configDir . '/config.xml'),
        'SimpleXMLElement',
        LIBXML_NOBLANKS | LIBXML_NONET
    );
    if ($fresh === false) {
        throw new RuntimeException('The current configuration could not be read.');
    }
    $source = dom_import_simplexml($fresh);
    $destination = dom_import_simplexml($config->object());
    while ($destination->firstChild !== null) {
        $destination->removeChild($destination->firstChild);
    }
    foreach ($source->childNodes as $child) {
        $destination->appendChild($destination->ownerDocument->importNode($child, true));
    }
}


const SPEEDTEST_SECTION = '//OPNsense/Speedtest/backup';

function speedtest_stored_section(): array
{
    /* Read the raw document rather than the model: a section that was never
       written has to stay distinguishable from one whose fields are all empty,
       and a field a later version added has to survive being read by this one. */
    $found = Config::getInstance()->object()->xpath(SPEEDTEST_SECTION);
    if (empty($found)) {
        return [];
    }
    $section = [];
    foreach ($found[0]->children() as $name => $value) {
        if ($value->count() === 0) {
            $section[(string)$name] = (string)$value;
        }
    }
    return $section;
}

function speedtest_shim_reply(array $value): void
{
    /* An object even when it is empty: the caller reads {} as "nothing stored". */
    echo json_encode((object)$value, JSON_UNESCAPED_SLASHES | JSON_UNESCAPED_UNICODE) . "\n";
}

$locked = false;
$exitCode = 0;
try {
    $verb = $argv[1] ?? '';
    if ($verb === 'import') {
        speedtest_shim_reply(speedtest_stored_section());
    } elseif ($verb === 'export') {
        $payload = json_decode((string)stream_get_contents(STDIN), true);
        if (!is_array($payload)) {
            throw new InvalidArgumentException('Expected one JSON object on standard input.');
        }
        /* Lock first: reading, changing and writing the configuration is one
           step, and another process may be saving its own section meanwhile. */
        backup_lock_current(Config::getInstance());
        $locked = true;
        $stored = speedtest_stored_section();
        if (array_key_exists('_expected', $payload)) {
            $expected = $payload['_expected'];
            unset($payload['_expected']);
            $fields = $stored;
            ksort($fields, SORT_STRING);
            $encoded = json_encode((object)$fields, JSON_UNESCAPED_SLASHES | JSON_THROW_ON_ERROR);
            $revision = hash('sha256', str_replace("\x7f", '\\u007f', $encoded));
            if (!is_string($expected) || !preg_match('/^[a-f0-9]{64}$/D', $expected) ||
                !hash_equals($revision, $expected)) {
                throw new RuntimeException('The configuration changed before the backup was saved.');
            }
        }
        $model = new Backup();
        $updates = [];
        foreach ($payload as $name => $value) {
            $field = $model->getNodeByReference((string)$name);
            if ($field !== null && !$field->isContainer() && is_scalar($value)) {
                $text = (string)$value;
                if (preg_match('/[^\x{0009}\x{000A}\x{000D}\x{0020}-\x{D7FF}\x{E000}-\x{FFFD}\x{10000}-\x{10FFFF}]/u', $text)) {
                    throw new InvalidArgumentException('A backup value cannot be represented in XML.');
                }
                $updates[(string)$name] = $text;
            }
        }
        $changed = false;
        foreach ($updates as $name => $value) {
            if (!array_key_exists($name, $stored) || $stored[$name] !== $value) {
                $changed = true;
            }
        }
        if ($changed) {
            /* Update named fields only; future fields and node attributes belong
               to the restored configuration and survive an older plugin save. */
            $root = Config::getInstance()->object();
            if (!isset($root->OPNsense)) {
                $root->addChild('OPNsense');
            }
            if (!isset($root->OPNsense->Speedtest)) {
                $root->OPNsense->addChild('Speedtest');
            }
            if (!isset($root->OPNsense->Speedtest->backup)) {
                $root->OPNsense->Speedtest->addChild('backup');
            }
            $parent = dom_import_simplexml($root->OPNsense->Speedtest->backup);
            foreach ($updates as $name => $value) {
                $matches = [];
                foreach ($parent->childNodes as $child) {
                    if ($child instanceof DOMElement && $child->nodeName === $name) {
                        $matches[] = $child;
                    }
                }
                $target = $matches[0] ?? $parent->appendChild($parent->ownerDocument->createElement($name));
                while ($target->firstChild !== null) {
                    $target->removeChild($target->firstChild);
                }
                $target->appendChild($parent->ownerDocument->createTextNode($value));
                foreach (array_slice($matches, 1) as $duplicate) {
                    $parent->removeChild($duplicate);
                }
            }
            Config::getInstance()->save(['description' => 'speedtest settings mirrored into the configuration']);
        }
        Config::getInstance()->unlock();
        $locked = false;
        speedtest_shim_reply(['changed' => $changed]);
    } else {
        throw new InvalidArgumentException('Usage: config_mirror.php export|import');
    }
} catch (Throwable $error) {
    speedtest_shim_reply(['status' => 'failed', 'error' => 'The native Speedtest configuration backup operation failed.']);
    $exitCode = 1;
} finally {
    if ($locked) {
        Config::getInstance()->unlock();
    }
}
exit($exitCode);

File: src/os-speedtest/src/usr/local/opnsense/scripts/speedtest/config_mirror.py
#!/usr/local/bin/python3
"""Decide what of the speed test survives a rebuild, and put it back.

An OPNsense backup is /conf/config.xml and nothing else. The page keeps its
state in /var/db/speedtest, so a restored box comes up with none of it: the
three choices an operator actually made -- which interface to leave from, which
server to ask, how many connections to open -- are gone, while everything else
in that directory (the last result, a run's progress, the cached server lists,
the run lock) is a measurement the plugin makes again on first use and would be
wrong to carry anyway. Only the three are mirrored.

The file stays the source of truth that api.php reads and writes; config.xml is
a one-directional copy that the native backup then carries for free. The two
functions below are the whole policy, which is why they are here and not in
config_mirror.php: they read and write nothing but this plugin's own files and
can therefore be tested without a router.
"""
import hashlib
import json
import os
import re
import subprocess
import sys
import tempfile

# Rooted for the suite; empty on a router, where these are absolute paths.
STATE = '/var/db/speedtest'
SETTINGS = STATE + '/settings.json'
SHIM = '/usr/local/opnsense/scripts/speedtest/config_mirror.php'
MARKER = '/conf/os-config-backup/Speedtest/applied.sha256'
PHP = '/usr/local/bin/php'
# The shim loads the OPNsense model stack, so it is slow rather than quick to
# fail; long enough for that, short enough that a settings save never hangs.
TIMEOUT = 20

FIELDS = ('interface', 'server_id', 'threads')
# The same defaults api.php answers with when the file is absent. A restore has
# to produce the settings the page would have shown, not an empty form.
DEFAULTS = {'interface': 'auto', 'server_id': '', 'threads': '4'}
# What api.php would accept from the page, applied in both directions: nothing
# this plugin would refuse is written to the configuration, and nothing the
# configuration carries is written to disk unless the page could have set it.
INTERFACE = re.compile(r'^[A-Za-z0-9_.:-]{1,32}$')
SERVER_ID = re.compile(r'^[0-9]{0,12}$')
THREADS = (1, 16)


class MirrorError(Exception):
    """The configuration could not be reached. Never the caller's problem."""


def root():
    return os.environ.get('OS_SPEEDTEST_ROOT', '')


def state_dir():
    return root() + STATE


def settings_file():
    return root() + SETTINGS


def shim():
    return root() + SHIM


def field_value(name, raw):
    """The value as this version would store it, or None if it would not."""
    if isinstance(raw, bool) or raw is None or isinstance(raw, (dict, list)):
        return None
    if name == 'threads':
        try:
            number = int(str(raw).strip())
        except (TypeError, ValueError):
            return None
        return str(number) if THREADS[0] <= number <= THREADS[1] else None
    text = str(raw).strip()
    if name == 'interface':
        return text if INTERFACE.match(text) else None
    return text if SERVER_ID.match(text) else None


def read_settings():
    """What the page last wrote, or nothing at all if it never did."""
    try:
        with open(settings_file()) as handle:
            stored = json.load(handle)
    except (OSError, ValueError):
        return {}
    return {str(name): value for name, value in stored.items()} if isinstance(stored, dict) else {}


def write_settings(settings):
    """Replace the file the way api.php does, so the page never reads a half."""
    directory = state_dir()
    os.makedirs(directory, mode=0o700, exist_ok=True)
    os.chmod(directory, 0o700)
    handle, temporary = tempfile.mkstemp(dir=directory, prefix='.settings.')
    try:
        with os.fdopen(handle, 'w') as out:
            json.dump(settings, out, separators=(',', ':'))
        os.chmod(temporary, 0o600)
        os.replace(temporary, settings_file())
    finally:
        if os.path.exists(temporary):
            os.unlink(temporary)


def mirror_payload():
    """The three settings, as the configuration should carry them.

    A value the page could not have produced is replaced by the default rather
    than copied: the configuration is what a rebuilt box will be given back, so
    it may only ever hold something this plugin would accept.
    """
    stored = read_settings()
    payload = {}
    for name in FIELDS:
        value = field_value(name, stored.get(name))
        payload[name] = DEFAULTS[name] if value is None else value
    return payload


def adopt_payload(payload):
    """Write the on-disk settings a restored configuration asked for.

    Only names this version knows and values it would accept are taken, so a
    section that is absent, half written, newer than this plugin or simply
    wrong leaves the rest of the file as it was. Nothing is written when
    nothing moved, which is what makes running this on every boot harmless.
    """
    if not isinstance(payload, dict):
        payload = {}
    stored = read_settings()
    wanted = dict(stored)
    for name in FIELDS:
        value = field_value(name, payload.get(name))
        if value is not None:
            wanted[name] = value
    if wanted == stored:
        return False
    write_settings(wanted)
    return True


def run_shim(verb, payload=None):
    """Hand the configuration half to the one language that can do it."""
    try:
        done = subprocess.run([PHP, shim(), verb],
                              input='' if payload is None else json.dumps(payload),
                              capture_output=True, text=True, timeout=TIMEOUT)
    except (OSError, subprocess.SubprocessError) as error:
        raise MirrorError('%s %s: %s' % (PHP, verb, error))
    if done.returncode != 0:
        raise MirrorError('%s exited %d: %s' % (verb, done.returncode,
                                                (done.stdout + done.stderr).strip()))
    try:
        answer = json.loads(done.stdout or '{}')
    except ValueError:
        raise MirrorError('%s answered with something other than JSON' % verb)
    if not isinstance(answer, dict):
        raise MirrorError('%s answered with something other than an object' % verb)
    return answer


def mirror():
    """Copy the settings on disk into the configuration."""
    stored = run_shim('import')
    marker = applied_backup()
    if any(name in stored for name in FIELDS) and marker and marker != revision_token(stored):
        raise MirrorError('A restored Speedtest configuration is pending; import it before saving settings.')
    payload = mirror_payload()
    payload['_expected'] = revision_token(stored)
    result = run_shim('export', payload)
    confirmed = run_shim('import')
    expected = dict(stored, **{name: payload[name] for name in FIELDS})
    if confirmed != expected:
        raise MirrorError('The Speedtest backup changed before its update could be confirmed.')
    mark_backup(confirmed)
    return result


def revision_token(fields):
    return hashlib.sha256(json.dumps(fields, ensure_ascii=True, sort_keys=True,
                                    separators=(',', ':')).encode()).hexdigest()


def applied_backup():
    try:
        with open(root() + MARKER, encoding='ascii') as handle:
            value = handle.read(65).strip()
        return value if re.fullmatch(r'[a-f0-9]{64}', value) else None
    except (OSError, UnicodeError):
        return None


def mark_backup(fields):
    path = root() + MARKER
    directory = os.path.dirname(path)
    os.makedirs(directory, mode=0o700, exist_ok=True)
    fd, temporary = tempfile.mkstemp(prefix='.backup-applied.', dir=directory)
    try:
        with os.fdopen(fd, 'w', encoding='ascii') as handle:
            handle.write(revision_token(fields) + '\n')
            handle.flush()
            os.fsync(handle.fileno())
        os.replace(temporary, path)
    finally:
        if os.path.exists(temporary):
            os.unlink(temporary)


def import_config():
    """Take the settings a restored configuration carries back to disk."""
    payload = run_shim('import')
    if not any(payload.get(name, '') != '' for name in FIELDS):
        return False
    if not any(field_value(name, payload.get(name)) is not None for name in FIELDS):
        return False
    # Failed mirror writes leave the last XML revision marked as applied. A
    # routine boot must keep the newer file; an erased settings store still
    # needs to be reconstructed after a rebuild or a /var reset.
    if applied_backup() == revision_token(payload) and os.path.isfile(settings_file()):
        return False
    changed = adopt_payload(payload)
    if run_shim('import') != payload:
        raise MirrorError('The Speedtest backup changed during restoration; retry the import.')
    mark_backup(payload)
    return changed


def main(argv):
    verb = argv[0] if argv else ''
    try:
        if verb == 'mirror':
            print(json.dumps(mirror()))
        elif verb == 'import-config':
            print(json.dumps({'changed': import_config()}))
        else:
            sys.stderr.write('usage: config_mirror.py mirror|import-config\n')
            return 2
    except (MirrorError, OSError) as error:
        # The settings are on disk either way, so this is reported and not
        # raised; both callers treat the exit status as a warning.
        sys.stderr.write('%s\n' % error)
        print(json.dumps({'changed': False, 'error': str(error)}))
        return 1
    return 0


if __name__ == '__main__':
    sys.exit(main(sys.argv[1:]))
